File: Back/NodeJS/src/application/features/retrieve-bets/RetrieveBetsQueryHandler.ts
import { IRequestHandler } from '../../Abstractions/Request/IRequestHandler';
import { IRequest } from '../../Abstractions/Request/IRequest';
import { IUserContext } from '../../Abstractions/IUserContext';
export class RetrieveBetsQueryHandler implements IRequestHandler<RetrieveBetsQuery, RetrieveBetsResponse[]> {
    constructor(private readonly dataAccess: IRetrieveBetsDataAccess,
        private userContext: IUserContext){}
    Handle(request: RetrieveBetsQuery): Promise<RetrieveBetsResponse[]> {
        return this.dataAccess.RetrieveBetsAsync(this.userContext.UserId);
    }
    GetRequestType(): string {
        return RetrieveBetsQuery.name;
    }
}

export class RetrieveBetsQuery implements IRequest<RetrieveBetsResponse[]> {
    constructor(){}
    Name: string = RetrieveBetsQuery.name;
}

export interface RetrieveBetsResponse {
    Id: string;
    Description: string;
    Coins: number;
    EndDate: Date;
    BookieId: string;
    BookieName: string;
    MaxAnswerDate: Date;
    Gamblers: GamblerDto[]
}

export interface IRetrieveBetsDataAccess {
    RetrieveBetsAsync(userId: string): Promise<RetrieveBetsResponse[]>;
}

export interface GamblerDto {
    Id: string;
    Name: string;
    HasAccepted?: boolean;
}

File: Back/.Net/BetFriends.Application/Features/RetrieveBets/RetrieveBetsQueryHandler.cs
using BetFriends.Application.Abstractions;
using BetFriends.Application.Abstractions.Messaging;

namespace BetFriends.Application.Features.RetrieveBets;

public class RetrieveBetsQueryHandler(IRetrieveBetsDataAccess retrieveBetsDataAccess,
                                      IUserContext userContext) : IQueryHandler<RetrieveBetsQuery, IEnumerable<RetrieveBetsResponse>>
{
    private readonly IRetrieveBetsDataAccess retrieveBetsDataAccess = retrieveBetsDataAccess;
    private readonly IUserContext userContext = userContext;

    public Task<IEnumerable<RetrieveBetsResponse>> Handle(RetrieveBetsQuery request, CancellationToken cancellationToken)
    {
        return retrieveBetsDataAccess.GetAsync(userContext.UserId);
    }
}


public record RetrieveBetsQuery : IQuery<IEnumerable<RetrieveBetsResponse>>;
public record RetrieveBetsResponse(Guid Id,
                                   string Description,
                                   DateTime EndDate,
                                   int Coins,
                                   Guid OwnerId,
                                   string OwnerName,
                                   DateTime MaxAnswerDate,
                                   IEnumerable<GamblerDto> Gamblers);

public record GamblerDto(Guid Id, string Name, bool? HasAccepted);

public interface IRetrieveBetsDataAccess
{
    Task<IEnumerable<RetrieveBetsResponse>> GetAsync(Guid memberId);
}